File: lab04/calculator/lib/Calculator.ex
defmodule Calculator do
  @moduledoc """
  Simple calculator implementation.
  """

  @doc """
  Adds two numbers.
  """
  def add(a, b), do: a + b

  @doc """
  Subtracts two numbers.
  """
  def subtract(a, b), do: a - b

  @doc """
  Multiplies two numbers.
  """
  def multiply(a, b), do: a * b

  @doc """
  Divides two numbers.
  """
  def divide(a, b) when b != 0, do: a / b
  def divide(_, 0), do: {:error, "Division by zero"}

  @doc """
  Squares a number.
  """
  def square(a), do: a * a

  @doc """
  Computes the square root of a number.
  """
  def square_root(a) when a >= 0, do: :math.sqrt(a)
  def square_root(_), do: {:error, "Cannot compute square root of a negative number"}
end

File: lab09/micro_benchmark/src/main/java/com/example/calculator/Calculator.java
package com.example.calculator;

/**
 * A simple calculator class that provides basic arithmetic operations.
 */
public class Calculator {

    /**
     * Adds two integers and returns the result.
     *
     * @param a the first integer
     * @param b the second integer
     * @return the sum of a and b
     */
    public int add(int a, int b) {
        return a + b;
    }

    /**
     * Subtracts the second integer from the first and returns the result.
     *
     * @param a the first integer
     * @param b the second integer
     * @return the difference between a and b
     */
    public int subtract(int a, int b) {
        return a - b;
    }

    /**
     * Multiplies two integers and returns the result.
     *
     * @param a the first integer
     * @param b the second integer
     * @return the product of a and b
     */
    public int multiply(int a, int b) {
        return a * b;
    }

    /**
     * Divides the first integer by the second and returns the result.
     * If the divisor is zero, an IllegalArgumentException is thrown.
     *
     * @param a the first integer
     * @param b the second integer
     * @return the quotient of a and b
     * @throws IllegalArgumentException if b is zero
     */
    public double divide(int a, int b) {
        if (b == 0) {
            throw new IllegalArgumentException("Division by zero is not allowed.");
        }
        return (double) a / b;
    }
}
